File: src/turtle_package/src/data_scope_pub.cpp
#include <ros/ros.h>
#include <geometry_msgs/Twist.h>
#include <turtlesim/Pose.h>
#include <cmath>
#include <std_msgs/Float64.h> 

// 全局变量存储海龟的位置和朝向
turtlesim::Pose group4_pose;
turtlesim::Pose group4_2_pose;

// 回调函数更新Group4的位置
void group4PoseCallback(const turtlesim::Pose::ConstPtr& msg) {
    group4_pose = *msg;
}

// 回调函数更新Group4_2的位置和朝向
void group4_2PoseCallback(const turtlesim::Pose::ConstPtr& msg) {
    group4_2_pose = *msg;
}

// 计算两只海龟之间的直线距离
double calculateDistance(const turtlesim::Pose& pose1, const turtlesim::Pose& pose2) {
    return sqrt(pow(pose1.x - pose2.x, 2) + pow(pose1.y - pose2.y, 2));
}

int main(int argc, char **argv) {
    // 初始化ROS节点
    ros::init(argc, argv, "data_scope_pub");
    ros::NodeHandle nh;

    // 订阅两只海龟的位置话题
    ros::Subscriber group4_sub = nh.subscribe("/Group4/pose", 10, group4PoseCallback);
    ros::Subscriber group4_2_sub = nh.subscribe("/Group4_2/pose", 10, group4_2PoseCallback);

    // 创建发布者发布距离和朝向角度
    ros::Publisher distance_pub = nh.advertise<std_msgs::Float64>("/turtle_distance", 10);
    ros::Publisher orientation_pub = nh.advertise<std_msgs::Float64>("/turtle_2_orientation", 10);

    ros::Rate loop_rate(10); // 10 Hz

    while (ros::ok()) {
        ros::spinOnce();

        // 计算距离并发布
        std_msgs::Float64 distance_msg;
        distance_msg.data = calculateDistance(group4_pose, group4_2_pose);
        distance_pub.publish(distance_msg);

        // 发布Group4_2的朝向角度
        std_msgs::Float64 orientation_msg;
        orientation_msg.data = group4_2_pose.theta;
        orientation_pub.publish(orientation_msg);

        // 打印距离和角度
        ROS_INFO("Group4_Pose - X: %f, Y: %f, Theta: %f    Distance: %f, Group4_2 theta: %f", group4_pose.x, group4_pose.y, group4_pose.theta, distance_msg.data, group4_2_pose.theta);
        loop_rate.sleep();
    }

    return 0;
}

File: src/turtle_py/scripts/data_scope_pub.py
#!/usr/bin/env python
import rospy
from turtlesim.msg import Pose
from std_msgs.msg import Float64
import math

group4_pose = Pose()
group4_2_pose = Pose()

def group4_pose_callback(msg):
    global group4_pose
    group4_pose = msg

def group4_2_pose_callback(msg):
    global group4_2_pose
    group4_2_pose = msg

def calculate_distance(pose1, pose2):
    return math.sqrt((pose1.x - pose2.x)**2 + (pose1.y - pose2.y)**2)

def main():
    rospy.init_node('data_scope_pub')
    rospy.Subscriber('/Group4/pose', Pose, group4_pose_callback)
    rospy.Subscriber('/Group4_2/pose', Pose, group4_2_pose_callback)
    distance_pub = rospy.Publisher('/turtle_distance', Float64, queue_size=10)
    orientation_pub = rospy.Publisher('/turtle_2_orientation', Float64, queue_size=10)
    rate = rospy.Rate(10)

    while not rospy.is_shutdown():
        distance_msg = Float64()
        distance_msg.data = calculate_distance(group4_pose, group4_2_pose)
        distance_pub.publish(distance_msg)

        orientation_msg = Float64()
        orientation_msg.data = group4_2_pose.theta
        orientation_pub.publish(orientation_msg)

        rospy.loginfo("Group4_Pose - X: %f, Y: %f, Theta: %f    Distance: %f, Group4_2 theta: %f",
                      group4_pose.x, group4_pose.y, group4_pose.theta, distance_msg.data, group4_2_pose.theta)
        rate.sleep()

if __name__ == '__main__':
    main()